README.md
# Robopager

Robopager is a job monitor and notification tool designed for daily email check and job latency check. The daily email check function can help you to monitor your inbox and whether you receive the expected emails at a certain time on a daily basis. The job latency check function helps to monitor every few seconds or minutes whether your scheduled workflows finish on time. If emails are not received in an expected timeframe or the scheduled jobs don't complete on time, you will receive the alerts from PagerDuty in email, text message or phone call depending on your choice. 

## Installation
Robopager requires Python 3.6+ __(need to comfirm)__
```
python3 -m venv <virtual environment name>
source <virtual environment name>/bin/activate
pip install robopager
```

## Setup Instructions
Before start using this tool, you need to configure the settings in the __etl.cfg__ and create your __checklist.yaml__ file. You can find example files in this folder and change to your own values.

### 1. etl.cfg

This file contains the loggin or server information of the applications or services that will be used in this tool, such as your monitored email, PagerDuty service and other servers. It also has the settings for you to choose the function mode.

* google_apps: enter the email address you want to monitor and use b64encode to encrypt your password for security concerns
* pager_duty: enter the information of the subdomain you will use and the API access key for authentication
* enable_redis: enter "True" if you want to use Redis, which is highly recommended because of API rate limit in PagerDuty; otherwise "False". Please refer to Note section for more information about Redis and please refer to this __[link](https://v2.developer.pagerduty.com/docs/rate-limiting)__ for more details about API rate limit.
* redis: enter the server and database information to get access to redis database, if you opt to use redis to store the job cache
* batchy: enter the server and port information to connect to your scheduled batch jobs
* function_type: you can enter either "email" for daily email check or "batchy" for intraday job latency check or both, separated by comma
* heartbeat: enter the "Integration Key" of the startup service created on PagerDuty. It usually consists of 32 characters. Since we recommend users to restart the robopager application at regular intervals to avoid the unnecessary crash-down of the checks, heartbeat as a service can notify user that robopager has restarted successfully
* timezone: your local timezone to make sure the correct delivery time. You can use the code below to search the correct timezone:
```
import pytz
for tz in pytz.all_timezones:
    print tz
```

### 2. checklist.yaml

This file lists the information of all the jobs that you want to monitor. Each job must have a unique name. You could find examples below for both types of job.

__Daily Email Checks__
* type: email
* pd_description: a brief description of the check job
* pd_service: the "Integration Key" of a PagerDuty service only for this specific job, usually consists of 32 characters
* senders: list of senders that robopager will be monitoring for particular emails
* delivery_time: the expected delivery time of emails you are monitoring; emails received before this time will not be scanned. Please pay attention to your local timezone
* check_time: the time at which robopager actually starts checking for delivery. Even if an email is expected to arrive at 7:00AM we may wait untill 7:15 to account for minor latency and avoid false alarms. Please pay attention to pass the time that is in the same timezone where you run the jobs, such as your local computer or cloud server, etc.
* subjects: a list of email subjects requires to satisify the check
```
unique_email_check_job_name:
    type: email
    pd_description: "Check whether you are OK"
    pd_service: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
    senders: ['email_address@gmail.com']
    delivery_time: 03:00
    check_time: 03:15
    subjects:
        - "Great - You look awesome!"
        - "Good - You look OK!"
```

__Intraday Latency Check__
* type: batchy
* pd_description: a brief description of the check job
* pd_service: the "Integration Key" of a PagerDuty service only for this specific job, usually consists of 32 characters
* wf_name: name of the scheduled workflow you want to monitor, please enter the correct and accurate name
* check_time: the time at which robopager actually starts checking
* poll_sec: the number of seconds in between checks
* latency_min: when it has been more than x minutes from last end time a failure will be generated, warnings at 80% of threshold
```
unique_latency_check_name:
    type: batchy
    pd_description: "intraday latency check for xxxx job"
    pd_service: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
    wf_name: "intraday_latency_check"
    check_time: "09:00"
    poll_sec: 180
    latency_min: 60
```

## Quick Start
After finish all the setups, you can start to use the tool to help monitor you scheduled jobs.
Robopager has two modes:
* Single Job Mode: This will run the specific check and exit when complete. Execute with a -c parameter and a specific job name you want to run. For example:
```
python3 -m robopager.robopager -c unique_email_check_job_name
```

* Service Mode: This will execute all the check jobs in the checklist.yaml file.
```
python3 -m robopager.robopager
```
In service mode, if you want to specify a certian checklist among several checklist file, you can:
```
python3 -m robopager.robopager -y checklist_name
```

## Notes
* __Redis__: The main purpose of using Redis is to prevent Robopager from being annoying. Using Redis allows PagerDuty to store the incident history data in it. PagerDuty will check the state of the last run in Redis and only create a new incident if the same key of last run is not found (key is deterministic based on date + hour). This will avoid PagerDuty from sending the same alerts in every run during the check period. For each check, two key patterns of a new incident will be created:

    * _jobname_: stores latest state for a check (success or failure)
    * _alert key_: stores pagerduty submission informaton
  
    Below are a few helpful redis commands:

    * connect to redis cli, assuming database 1 will be used for robopager: `redis-cli -n 1`
    * list keys based on pattern: `keys *` or `keys dw_Lockers*`
    * get all fields within a hash (Robopager stores all keys as hashes): `hgetall dw_Lockers_245am`

* This modules uses datacoco_core's config() to parse the configuration file. If you have preferable configuration tool, you can change to it and the corresponding parsing code in the beginning section of robopager.py
* Robopager submits to PagerDuty using a deterministic key, therefore you can have multiple Robopager instance running (for redundancy without producing duplicate tickets). There is a special offset_sec parameter in the PDInteraction class, setting this will have a specific server wait the specified number of seconds before checking state, and subsequently triggering incidents in Pagerduty.  This will prevent unnecessary API calls.
* Robopager will not yet reload the checklist.yaml if changed, it will need to be restarted
* Finally, this is not a very sophisticated application (this simplicity is deliberate), we are using features like threading and some 0.x modules. So, we suggest rebooting or restarting the service often (weekly or daily) to avoid the unnecessary crash-down of the checks

## Development
### Getting Started
It is recommended to use the steps below to set up a virtual environment for development:
```
python3 -m venv <virtual env name>
source <virtual env name>/bin/activate
pip install -r requirements.txt
```
### Testing
```
pip install -r requirements-dev.txt
```
To run the testing suite, simply run the command:
```
python3 -m unittest discover tests
```

## Contributing
Contributions to Robopager are welcome!
Please reference guidelines to help with setting up your development environment __here__


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1] - 2021-04-29
### Added
    - Fix pyyaml security severity

## [0.1.0] - 2020-03-17
### Added
    - Initial project setup

[0.1.0]: https://github.com/equinoxfitness/robopager/releases/tag/0.1.0
[0.1.1]: https://github.com/equinoxfitness/robopager/releases/tag/0.1.1

tests/test_robopager.py
import unittest
from robopager.robopager import parse_checklist, PDInteraction
from robopager.check_type.daily_email_check import CheckEmails
from robopager.check_type.intraday_latency_check import CheckWF
from robopager.check_type import intraday_latency_check
from datetime import datetime
import pytz
from unittest.mock import patch, MagicMock
from datacoco_core.config import config


class TestCommon(unittest.TestCase):
    def test_parse_checklist(self):
        """
        test reading of config yaml file and its output format
        :return:
        """

        config = "tests/test_data/test.yaml"
        expected = {
            "checklist_parse_test": {
                "type": "batchy",
                "pd_description": "Test for parsing checklist",
                "pd_service": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
                "wf_name": "checklist_parse_test",
                "check_time": "09:00",
                "poll_sec": 180,
                "latency_min": 60,
            }
        }
        found = parse_checklist(config)
        self.assertDictEqual(expected, found)


class TestBatchyWF(unittest.TestCase):
    def setUp(self):
        self.b = CheckWF("test_robopager", "0.0.0.0", "8050")

    @patch("datacoco_batch.batch.Batch.get_status")
    @patch.object(
        intraday_latency_check, "datetime", MagicMock(wraps=datetime)
    )
    def test_check_batchy_wf_success(self, mock_get_status):
        mock_get_status.return_value = {
            "global": {
                "batch_start": "2020-03-04T11:20:00.00",
                "status": "success",
            }
        }

        naive_time = datetime(2020, 3, 4, 11, 22, 0, 0)
        intraday_latency_check.datetime.now.return_value = pytz.utc.localize(
            naive_time
        )
        failure_count, result = self.b.check_batchy_wf(max_latency=3)
        self.assertEqual(failure_count, 0)
        self.assertEqual(result["alert_level"], "SUCCESS")


class TestPDInteraction(unittest.TestCase):
    def setUp(self):

        conf = config("tests/test_data/test_etl.cfg")
        sub_domain = conf["pager_duty"]["subdomain"]
        api_access_key = conf["pager_duty"]["api_access_key"]
        self.pd_service = conf["pager_duty"]["test_pd_service"]

        self.pd = PDInteraction(sub_domain, api_access_key)

    def test_trigger_incident(self):
        with patch("robopager.robopager.ri", create=True) as ri:
            resp = self.pd.trigger_incident(
                self.pd_service,
                "test_robopager",
                "test_robopager",
                check="test",
                override=True,
            )
            print("resp: {}".format(resp))
            self.assertEqual(resp, 200)


class TestCheckEmails(unittest.TestCase):
    def setUp(self):

        self.e = CheckEmails(
            "username",
            "password",
            "Successful - Job1 Completed",
            ["address@gmail.com"],
            "US/Easter",
        )
        self.e.subjects = ["Successful - Job1 Completed"]
        self.e.emails_received = [
            "Successful - Job2 Completed",
            "Successful - Job3 Completed",
            "Successful - Job4 Completed",
            "Successful - Job1 Completed",
        ]

        print(self.e.subjects)

    def test_check_missing_emails(self):

        failure_count, results = self.e.check_missing_emails()

        self.assertEqual(failure_count, 0)
        self.assertEqual(results, [{"Successful - Job1 Completed": "ok"}])

    def test_parse_gmail_dates(self):
        gdate = "Wed, 2 Jan 2019 01:47:40 -0500"
        date = self.e.parse_gmail_dates(gdate)
        self.assertEqual(date, datetime(2019, 1, 2, 1, 47, 40))


if __name__ == "__main__":
    unittest.main()


robopager/config_wrapper.py
import sys
from datacoco_core.config import config
from datacoco_secretsmanager import SecretsManager


class ConfigWrapper:
    """
    Wrapper file for config management for robopager
    """

    @staticmethod
    def extend_parser(parser):
        parser.add_argument(
            "-cf",
            "--config",
            required=True,
            default="core",
            help="""
                choose whether to use secret_manager or
                datacoco_core to retrieve credentials
                """,
            choices=["secret_manager", "core"],
        )

        if "core" in sys.argv:
            parser.add_argument(
                "-p",
                "--path",
                default="etl.cfg",
                help="""
            enter the path direct to the config file
            """,
            )

        return parser

    @staticmethod
    def process_config(args):
        if args.config == "secret_manager":
            conf = SecretsManager().get_config(
                project_name="maximilian", team_name="data"
            )
        elif args.config == "core":
            conf = config(args.path)

        return conf


robopager/check_type/daily_email_check.py
# Email check function

import imaplib
import email
import sys
from datetime import datetime, date
import pytz
from datacoco_core.logger import Logger


log = Logger()


def convert_local_to_utc(t, timezone):
    # convert local time to utc (both are naive)
    local = pytz.timezone(timezone)
    local_dt = local.localize(t)
    utc_dt = local_dt.astimezone(pytz.utc)
    naive_utc = utc_dt.replace(tzinfo=None)
    return naive_utc


class CheckEmails:
    """
    scans and filters gmail inbox
    """

    def __init__(
        self,
        username,
        password,
        subjects,
        senders,
        timezone,
        received_after="00:00",
    ):
        """
        :param subjects:
        :param senders:
        :param received_after:  default is midnight
        :return:
        """

        self.imapSession = imaplib.IMAP4_SSL("imap.gmail.com")
        self.username = username
        self.password = password

        self.subjects = subjects
        self.senders = senders
        self.received_after = received_after
        self.emails_received = []
        self.timezone = timezone

    def login(self):
        log.l("connecting to gmail")
        typ, accountDetails = self.imapSession.login(
            self.username, self.password
        )
        if typ != "OK":
            log.l("Not able to sign in!")
            sys.exit()

    def get_emails(self):
        """
        pulls down a list of emails, filters by time, subject, senders
        :return: emails recieved matching criteria
        """

        self.login()

        # select messages
        date_received = (date.today()).strftime("%d-%b-%Y")
        delivery_threshold = datetime.strptime(
            " ".join((date_received, self.received_after)), "%d-%b-%Y %H:%M"
        )
        utc_delivery_threshold = convert_local_to_utc(
            delivery_threshold, self.timezone
        )

        imap_search = '(SENTSINCE {date} FROM "{senders}")'.format(
            date=date_received, senders=",".join(self.senders)
        )
        self.imapSession.select('"[Gmail]/All Mail"')
        typ, data = self.imapSession.search(None, imap_search)
        if typ != "OK":
            log.l("Error searching Inbox.")
            sys.exit()
        log.l(str(len(data[0].split())) + " emails found")

        # Iterating over all emails
        for msgId in data[0].split():
            # typ, messageParts = imapSession.fetch(msgId, '(RFC822)')
            typ, messageParts = self.imapSession.fetch(
                msgId, "(BODY.PEEK[HEADER])"
            )
            msg = email.message_from_bytes(messageParts[0][1])
            # print msg['Subject']
            # print msg['Date']
            time_recieved = self.parse_gmail_dates(msg["Date"])
            if time_recieved < utc_delivery_threshold:
                continue
            self.emails_received.append(msg["Subject"].replace("\r\n", ""))
            if typ != "OK":
                log.l("Error fetching mail.")
                raise
                exit(1)
        log.l("emails recieved: " + ",".join(self.emails_received))
        return self.emails_received

    @staticmethod
    def parse_gmail_dates(gdate):
        try:
            date = datetime.strptime(gdate[:-6], "%d %b %Y %H:%M:%S")
            return date
        except Exception:
            pass

        try:
            date = datetime.strptime(gdate[:-6], "%a, %d %b %Y %H:%M:%S")
            return date
        except Exception:
            date = datetime.strptime(gdate[:-12], "%a, %d %b %Y %H:%M:%S")
            return date

    def check_missing_emails(self):
        """
        compares list of expected subjects to what is retrieved
        :return: failure_count, results
        """
        log.l("look for emails not received")
        results = []
        failure_count = 0
        for x in self.subjects:
            r = {}
            if x in self.emails_received:
                r[x] = "ok"
            else:
                r[x] = "not received"
                failure_count += 1
            results.append(r)

        return failure_count, results

    def close_session(self):
        """
        closes imap sesion
        """
        self.imapSession.close()
        self.imapSession.logout()
        log.l("disconnected from gmail")


robopager/check_type/intraday_latency_check.py
# Intraday latency check function

from datetime import datetime
import pytz
from datacoco_batch.batch import Batch
from datacoco_core.logger import Logger


log = Logger()


def convert_time(t):
    # convert naive datetime object to utc aware datetime
    utc = pytz.utc
    timetz = utc.localize(t)
    return timetz


class CheckWF:
    """
    Calls batchy endpoint to get job status.
    """

    def __init__(self, wf, batchy_server, batchy_port):

        self.b = Batch(wf, batchy_server, batchy_port)
        log.l("Checking wf: {}".format(wf))

    def check_batchy_wf(self, max_latency):

        status = self.b.get_status().get("global")
        if status:
            failure_count, result = self.calc_latency_tests(
                status, max_latency
            )
        else:
            raise ValueError("Could not find wf")
        return failure_count, result

    @staticmethod
    def calc_latency_tests(result, max_latency):
        """
        run business logic on result to create alerts
        :param result:
        :param max_latency:
        :return:
        """
        failure_count = 0

        # use batch start, not end
        batch_start = result.get("batch_start")
        latency = (
            datetime.now(pytz.utc)
            - convert_time(
                datetime.strptime(batch_start, "%Y-%m-%dT%H:%M:%S.%f")
            )
        ).seconds / 60

        if latency >= max_latency:
            log.l(
                "latency: {} is greater than max latency: {}".format(
                    latency, max_latency
                )
            )
            failure_count = 1
            result["alert_level"] = "FAILURE"
            result["alert_message"] = "latency issue"
        elif result["status"] == "failure":
            log.l("failure b/c of job failure")
            result["alert_level"] = "FAILURE"
            result["alert_message"] = "job failure"
        elif latency >= max_latency * 0.8:
            log.l(
                "latency: {} is greater than 80% of max latency: {}".format(
                    latency, max_latency
                )
            )
            result["alert_level"] = "WARNING"
            result["alert_message"] = "passed 80% of latency threshold"
        else:
            result["alert_level"] = "SUCCESS"
            log.l("Success")
        result["latency"] = latency
        return failure_count, result


robopager/robopager.py
#!/usr/bin/env python
from datetime import datetime
import yaml
import json
import requests
import argparse
import schedule
import time
import threading
import redis
from datacoco_core.logger import Logger
from robopager.config_wrapper import ConfigWrapper
from robopager.check_type.daily_email_check import CheckEmails
from robopager.check_type.intraday_latency_check import CheckWF


# -------------------------------------------------------------------------------------------------------------------

log = Logger()
KEY_PREFIX = "rp"

# -------------------------------------------------------------------------------------------------------------------

# common functions


def parse_checklist(check_file):
    """
    reads yaml file into python object
    :param check_file:
    :return: checklist
    """
    with open(check_file, "r") as checklist_yaml:
        check_list = yaml.load(checklist_yaml)
    return check_list


def json_serial(obj):
    """
    JSON serializer for objects not serializable by default json code"
    :param obj:
    :return:
    """
    if isinstance(obj, datetime):
        serial = obj.isoformat()
        return serial
    raise TypeError("Type not serializable")


def time_2_datetime(time):
    """
    turns time into datetime, basd on today's date
    :param time:
    :return:
    """
    today_time = datetime.combine(
        datetime.today(), datetime.strptime(time, "%H:%M").time()
    )
    return today_time


def str_to_bool(en):
    if en == "True":
        return True
    elif en == "False":
        return False
    else:
        raise ValueError("please enter correct boolean values to enable Redis")


# -------------------------------------------------------------------------------------------------------------------
# core robopager


class RedisInteraction:
    """
    Enable Redis to check and log
    """

    def __init__(self, redis_server, redis_db):
        # connect to redis
        self.rconn = redis.StrictRedis(
            host=redis_server, port=6379, db=redis_db
        )

    def check_last_run(self, incident, override=False, offset_sec=0):

        """
        check the incident key of last run
        :param incident:
        :param override:
        :param offset_sec:
        :return:
        """

        # we'll sleep by the offset seconds, assuming
        # we have a cluster of robopager servers for redundancy prevents
        # multiple robopager services from dispatching alerts at the same time
        log.l("offset: sleeping for %ss" % offset_sec)
        time.sleep(float(offset_sec))

        # first check to make sure it hasn't already been opened,
        # otherwise notifications will be annoying
        last_incident = None

        try:
            last_incident = self.rconn.hget(
                KEY_PREFIX + "." + incident, "incident_dt"
            )
            log.l("last incident time: " + str(last_incident))
        except Exception:
            pass
            log.l("issue connecting to redis")

        if last_incident is not None and not override:
            log.l("already a triggered incident")
            return

    def log_to_redis(self, key_name, detail):
        """
        log the new incident or check result to Redis
        :param key_name:
        :param detail:
        :return:
        """

        try:
            self.rconn.hmset(
                KEY_PREFIX + "." + key_name,
                {"result": detail, "status_dt": str(datetime.now())},
            )
        except Exception:
            pass
            log.l("issue logging to redis")


class NonRedisInteraction:
    """
    Disable to check and log
    """

    def check_last_run(self, incident, override=False, offset_sec=0):
        log.l("Redis disabled, no need to check last run")

    def log_to_redis(self, key_name, detail):
        log.l("Redis disabled, no need to log to Redis")


class PDInteraction:
    """
    wrapper for pager duty
    """

    def __init__(self, sub_domain, api_access_key):

        self.sub_domain = sub_domain
        self.headers = {
            "Authorization": "Token token={0}".format(api_access_key),
            "Content-type": "application/json",
        }

    def trigger_incident(
        self,
        pd_service,
        pd_description,
        error_detail,
        check="eqx",
        override=False,
        offset_sec=0,
    ):
        """
        triggers a new pager duty incident
        :param pd_service:
        :param pd_description:
        :param error_detail:
        :param check:
        :param override:
        :param offset_sec:
        :return:
        """
        incident = check + "-" + str(datetime.now().strftime("%Y%m%d-%H"))

        # check the incident key of last run
        ri.check_last_run(incident, override, offset_sec)

        # if nothing is found we trigger the incident
        payload = json.dumps(
            {
                "service_key": pd_service,
                "incident_key": incident,
                "event_type": "trigger",
                "description": pd_description,
                "client": "robopager",
                "details": error_detail,
            }
        )
        p = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
        r = requests.post(p, headers=self.headers, data=payload,)

        if r.status_code != 200:
            log.l("Error triggering pager duty")
            raise
        else:
            # log the newly-triggered incident to Redis
            ri.log_to_redis(incident, payload)

        log.l(r.status_code)
        log.l(r.reason)
        log.l(r.text)
        return r.status_code


class JobRunner:
    """
    wrapper for schedule module
    """

    def __init__(self):
        """
        """

    def schedule_job(self, job, params, schedule_time, interval):
        """
        adds job to schedule
        :param job:
        :param params:
        :param schedule_time:
        :param interval: will run job at this interval in seconds.
        0 will be a check once daily
        """
        kw = {"func": job, "param": params}
        if interval == 0:
            schedule.every().day.at(schedule_time).do(self.run_threaded, **kw)
        else:
            # note scheduler does not have a start time concept
            # so this will need to be handled in the job
            schedule.every(interval).seconds.do(self.run_threaded, **kw)

    @staticmethod
    def run_threaded(func, param):
        """
        runs job within thread
        :param func:
        :param param:
        """
        # job_thread = threading.Thread(target=func, args=param)
        job_thread = threading.Thread(target=func, kwargs=param)
        job_thread.start()

    @staticmethod
    def dispatcher():
        """
        starts scheduler
        """
        last_hour = datetime.now().strftime("%H")
        print(last_hour)
        while True:
            schedule.run_pending()
            time.sleep(1)
            this_hour = datetime.now().strftime("%H")
            if this_hour != last_hour:
                log.l("heartbeat --> still here homey!")
                last_hour = this_hour

    @staticmethod
    def get_jobs():
        """
        returns list of jobs scheduled
        """
        return schedule.jobs


def run_test(**kwargs):
    """
    a wrapper for running basic tests
    :param kwargs:
    :return:
    """
    failure_count = 0
    # extract shared variables
    pd_check = kwargs.get("check")
    check_type = kwargs.get("type", "daily_email")
    pd_service = kwargs.get("pd_service")
    pd_description = kwargs.get("pd_description")
    check_time = kwargs.get("check_time")

    # now run the appropriate check based on check type
    if check_type == "email":
        subjects = kwargs.get("subjects")
        senders = kwargs.get("senders")
        delivery_time = kwargs.get("delivery_time")
        e = CheckEmails(
            username, password, subjects, senders, timezone, delivery_time
        )
        e.get_emails()
        failure_count, results = e.check_missing_emails()
        e.close_session()

    elif check_type == "batchy":
        # since schedule module doesn't have a start time concept,
        # we'll implement here
        if datetime.now() > time_2_datetime(check_time):
            latency_min = kwargs.get("latency_min")
            pd_description = kwargs.get("pd_description")
            wf_name = kwargs.get("wf_name")
            wf = CheckWF(wf_name, batchy_server, batchy_port)
            failure_count, results = wf.check_batchy_wf(latency_min)
        else:
            return

    else:
        log.l("Invalid test type, please check yaml")
        return

    # log check result to redis
    try:
        ri.log_to_redis(pd_check, json.dumps(results, default=json_serial))
    except Exception:
        log.l("Error logging roboapger check result to Redis")

    # logging and pd alerts
    if check_type == "email" or failure_count != 0:
        log.l(
            "failure count: "
            + str(failure_count)
            + "\nresults:"
            + str(results)
        )

    if failure_count > 0:
        log.l("!!!!uh oh bro!!!!")
        PDInteraction(sub_domain, api_access_key).trigger_incident(
            pd_service,
            pd_description,
            json.dumps(
                {
                    "results": results,
                    "sender": str(senders),
                    "check_time": check_time,
                    "delivery_time": delivery_time,
                },
                default=json_serial,
                indent=4,
            ),
            pd_check,
        )
    else:
        return


if __name__ == "__main__":
    # parse args
    log.l("starting")
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-c",
        "--check",
        help="enter a check referencing the yaml",
        default="cron",
    )
    parser.add_argument(
        "-y",
        "--checklist",
        help="enter a checklist file, default is checklist.yaml",
        default="cfg/checklist.yaml",
    )
    # add config parameter
    parser = ConfigWrapper.extend_parser(parser)

    args = parser.parse_args()
    check = args.check
    checklist = args.checklist

    conf = ConfigWrapper.process_config(args)

    # heartbeat alert
    heartbeat = conf["robopager"]["heartbeat_service"]

    # pd vars
    sub_domain = conf["pager_duty"]["subdomain"]
    api_access_key = conf["pager_duty"]["api_access_key"]

    # gmail vars
    username = conf["google_apps"]["email"]
    password = conf["google_apps"]["password"]

    # function type
    function_type = conf["function_type"]["type"]
    function_list = function_type.split(",")

    # batchy vars
    batchy_server = conf["batchy"]["server"]
    batchy_port = conf["batchy"]["port"]

    # timezone
    timezone = conf["timezone"]["timezone"]

    # enable redis
    enable_redis = conf["enable_redis"]["enable"]

    # enable & connect to redis
    print("\nRedis Enable Status: {}".format(enable_redis))

    if str_to_bool(enable_redis):
        redis_server = conf["redis"]["server"]
        redis_db = conf["redis"]["db"]
        # enable redis
        ri = RedisInteraction(redis_server, redis_db)
        log.l("Redis enabled")
    else:
        ri = NonRedisInteraction()
        log.l("Redis disabled")

    log.l("\n\nCHECKLIST: {}".format(checklist))

    # get list of tests and metadata
    check_list = parse_checklist(checklist)

    # check to see if we are running a specific check
    # via external cron or starting as a service
    if check != "cron":
        log.l("single job mode")
        if check not in check_list:
            log.l("invalid test, please review checklist.yaml")
            raise
        else:
            kw = check_list[check]
            kw["check"] = check
            run_test(**kw)
            exit()
    else:
        log.l("service mode")

    # create job runner instance
    j = JobRunner()
    log.l("scheduled job type:" + function_type)

    # loop
    for c in check_list:
        # only schedule jobs of enabled function types
        check_type = check_list[c].get("type")
        if check_type not in function_list:
            continue
        # only schedule jobs with valid check_time
        check_time = check_list[c].get("check_time")
        if len(str(check_time)) != 5:
            continue
        poll_interval = check_list[c].get("poll_sec", 0)
        check_params = check_list[c]
        check_params["check"] = c
        j.schedule_job(run_test, check_params, check_time, poll_interval)

    jobs = sorted(j.get_jobs())

    startup_msg = (
        str(len(jobs))
        + " of "
        + str(len(check_list))
        + " have been scheduled:\n-"
        + "\n- ".join(str(v) for v in jobs)
    )
    startup_status = "OK" if len(jobs) == len(check_list) else "FAILURE"

    log.l(startup_msg)
    PDInteraction(sub_domain, api_access_key).trigger_incident(
        heartbeat,
        startup_status + ": Robopager Startup",
        startup_msg,
        override=True,
    )
    j.dispatcher()
